File: algorithms/leetcode/median-of-two-arrays/median_of_two_arrays.py
"""
Pseudocode:
    1. Merge the two arrays
    2. Sort the merged array
    3. If the length of the merged array is even, return the average of the two middle elements
    4. If the length of the merged array is odd, return the middle element

    C <- A + B
    sort(C)
    loc <- floor(|C|/2)
    
    if mod(|C|, 2) == 0:
        return (C[loc] + C[loc - 1]) / 2
    else:
        return C[loc]

Time complexity:
    O(nlogn) where n is the total number of elements in the two arrays

Space complexity:
    O(n) where n is the total number of elements in the two arrays
"""
import unittest, sys
class Solution(unittest.TestCase):
    def self_soln(self, nums1: list[int], nums2: list[int]) -> float:
        """
        Given two sorted arrays nums1 and nums2 of size m and n respectively, return the median of the two sorted arrays.

        Args:
            nums1 (list): the first sorted array
            nums2 (list): the second sorted array

        Return:
            float: the median of the two sorted arrays
        
        Time complexity:
            O(nlogn) where n is the total number of elements in the two arrays

        Space complexity:
            O(n) where n is the total number of elements in the two arrays
        """
        #merge and re-sort
        nums = nums1 + nums2
        nums.sort()
        
        #length of array
        n = len(nums)
        i = n//2

        #even
        if n % 2 == 0:
            return(nums[i] + nums[i - 1]) / 2
        else:
            return nums[i]
    def test_1(self):
        self.assertEqual(self.self_soln([1,3], [2]), 2.0)
    def test_2(self):
        self.assertEqual(self.self_soln([1,2], [3,4]), 2.5)
    def test_3(self):
        self.assertEqual(self.self_soln([0,0], [0,0]), 0.0)
    def test_4(self):
        self.assertEqual(self.self_soln([], [1]), 1.0)

if __name__ == "__main__":
    unittest.main()
        

File: algorithms/leetcode/median-of-two-arrays/median_of_two_arrays.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <algorithm>
using namespace std;
class Solution{
public:
    double brute_force(vector<int>& nums1, vector<int>& nums2) {
        /*
        @brief brute force solution to find the median of two sorted arrays
        @param nums1: vector of integers
        @param nums2: vector of integers
        @return median of the two arrays

        @example
        nums1 = [1, 3]
        nums2 = [2]
        median = 2.0

        nums1 = [1, 2]
        nums2 = [3, 4]
        median = 2.5

        nums1 = [0, 0]
        nums2 = [0, 0]
        median = 0.0

        @note
        time: O(nlogn)
        space: O(n)

        @details
        The idea is to merge the two arrays and then sort them. Then, we can find the median by checking if the length of the merged array is odd or even
        */
        //array lengths
        int n = nums1.size();
        int m = nums2.size();
        //merge
        int i{0}, j{0};
        vector<int> merged;
        while(i < n || j < m) {
            if (i < n) {
                merged.push_back(nums1[i++]);
            } else if (j < m){
                merged.push_back(nums2[j++]);
            }
        }
        //sort
        sort(merged.begin(), merged.end());
        
        if (merged.size() % 2 == 1) {
            return merged[merged.size() / 2];
        } else {
            return (merged[merged.size() / 2] + merged[merged.size() / 2 - 1]) / 2.0;
        }
    }
    double binary_search(vector<int>& nums1, vector<int>& nums2) {
        /*
        @brief binary search solution to find the median of two sorted arrays
        @param nums1: vector of integers
        @param nums2: vector of integers
        @return median of the two arrays

        @example
        nums1 = [1, 3]
        nums2 = [2]
        median = 2.0

        nums1 = [1, 2]
        nums2 = [3, 4]
        median = 2.5

        @note
        time: O(logn)
        space: O(1)

        @details
        The idea is to partition the two arrays such that the left partition is less than the right partition
        */
        int n=nums1.size(), m=nums2.size();
        if (n > m) return binary_search(nums2, nums1);
        int total{n + m}, left{(n + m + 1) / 2}, low{0}, high{n};
        while(low <= high) {
            int mid_num1{(low + high) >> 1}, mid_num2{(left - mid_num1)};
            auto l1{INT_MIN}, l2{INT_MIN}, r1{INT_MAX}, r2{INT_MAX};
            if (mid_num1 < n) r1 = nums1.at(mid_num1);
            if (mid_num2 < m) r2 = nums2.at(mid_num2);
            if (mid_num1 - 1 >= 0) l1 = nums1.at(mid_num1 - 1);
            if (mid_num2 - 1 >= 0) l2 = nums2.at(mid_num2 - 1);

            if (l1 <= r2 && l2 <= r1) {
                if (total % 2 == 1) {
                    return static_cast<double>(max(l1, l2));
                } else {
                    return static_cast<double>(max(l1, l2) + min(r1, r2)) / 2.0;
                }
            } else if (l1 > r2) {
                high = mid_num1 - 1;
            } else {
                low = mid_num1 + 1;
            }
        }
        return 0;
    }
};

int main() {
    Solution s;
    vector<int> nums1{1, 3};
    vector<int> nums2{2};
    auto header{"(brute force, binary search)"};
    cout << header << " -> (" << s.brute_force(nums1, nums2) << ", " 
        << s.binary_search(nums1, nums2) << ")" << endl;

    nums1 = {1, 2};
    nums2 = {3, 4};
    cout << header << " -> (" << s.brute_force(nums1, nums2) << ", " 
        << s.binary_search(nums1, nums2) << ")" << endl;

    nums1 = {0, 0};
    nums2 = {0, 0};
    cout << header << " -> (" << s.brute_force(nums1, nums2) << ", " 
        << s.binary_search(nums1, nums2) << ")" << endl;

    return 0;
}